### PR TITLE
#885 media viewer

### PR DESCRIPTION
**Pull request by dsipos-dev**
## References
* Fixes [#885](https://github.com/DSpace/dspace-angular/issues/885)

## Description
Creating a media viewer for image, video and audio files, if there is no difference between the file types in one item.

## Instructions for Reviewers
Replacing a thumbnail for a media viewer. We want to create an easier way to check media in dspace, rather than downloading them. Creating a media viewer for image, video and audio files, if there is no difference between the file types in one item.

* For images we want to add an image gallery, when we click on the thumbnail an image gallery pops up and we can view all images.
* For the videos and audios we want to create a video player with a playlist, if only one video is available in the item just replace the thumbnail with a video player, if there is more than one we add a playlist under the player.

Reviewers can test this PR by:
Test image:
* Create an item in a collection and add some images.
* Navigate to the item simple page view and on clicking the thumbnail and a gallery is going to popup.

Test Video:
* Create an item in a collection and add some videos.
* Navigate to the item simple page view and the thumbnail is replaced with a video player.
Change thumbnail for the video:
* Authenticating as an admin
* Going to Access Control → Edit → Item → Select current item
* Bitstream → Upload → Create thumbnail bundle
* Upload an image → Save
* Navigate back to item

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Comment by tdonohue** [comment]
Thanks for your contribution, @dsipos-dev !  I think this sort of feature would definitely be of interest to DSpace users.

I'll see if I can find someone to help review & provide advice on why your tests seem to all be failing (at least they are failing in our Travis CI checks in this PR).   @artlowel : would someone on your team have time to give this an initial quick review & provide advice for fixing the tests?

**Comment by kanasznagyzoltan** [comment]
@tdonohue We have found a forked version of ngx-gallery which is fully Angular 8 compatible. We are trying to change the gallery to that one.

**Comment by kanasznagyzoltan** [comment]
@tdonohue @artlowel 
We have written our first unit tests to the components belonging this issue (extension).

File: src/app/app.component.scss
Comment (artlowel) [inline src/app/app.component.scss:42]: Why remove the header navbar's z-index?
If it needs to change I'd prefer you change the variable value instead of removing it entirely

File: src/assets/i18n/en.json5
Comment (artlowel) [inline src/assets/i18n/en.json5:2467]: Small typo: this should be "Previous"

**Review by artlowel — CHANGES_REQUESTED** [review]
Thanks for the contribution @dsipos-dev!

The code looks good, so do the tests. The feature in general works well and looks good.

I have a few points of feedback though:
- Image gallery
	- If you have a bitstream in the original bundle that can't be downloaded for any reason (e.g. file not found, or your ORIGINAL bundle contains files that aren't supported by the gallery plugin, such as video), the gallery will stop responding when that bitstream is reached
	- If you're an authenticated user the left arrow button is shown underneath the sidebar. I'd expect it to leave space for the sidebar or if that's not easy to do, go over it.
- Video player
	- When you open an item page with the video player on it, the video will start to download immediately. I'd change it so the user has to interact with the player first. That way we don't use up unnecessary bandwidth for the user and the server. And we don't trigger "file download" stats unnecessarily
	- There's a problem with unsupported file types here as well, files in the ORIGINAL bundle that aren't videos will still show up in the playlist

In general, I'd also like to make the use of these features configurable. I'd add a section to environment.ts for mediaViewers, with booleans to enable the image gallery and video player separately

**Comment by dsipos-dev** [comment]
Thank for the review @artlowel

I had applied your requested changes:
- added environment
- filtered the unsupported media types
- and if user is authenticated, move left arrow to proper position

**Comment by kanasznagyzoltan** [comment]
Dear @artlowel, @dsipos-dev had applied your requested changes, could you inspect them?
For some reason the codecov status still waiting to be reported.

File: src/config/media-viewer-config.interface.ts
Comment (artlowel) [inline src/config/media-viewer-config.interface.ts:4]: Is there a reason to have a separate `enable` boolean? wouldn't simply setting the `image` or `video` boolean to true imply that the viewer is enabled?

File: src/environments/environment.common.ts
Comment (artlowel) [inline src/environments/environment.common.ts:223]: I'd disable the viewers by default

**Review by artlowel — COMMENTED** [review]
Thanks for taking my feedback in to account!

I have a few more minor comments inline

File: src/config/media-viewer-config.interface.ts
Comment (dsipos-dev) [reply src/config/media-viewer-config.interface.ts]: Yes its a bit redundant so i fixed it

**Comment by kanasznagyzoltan** [comment]
Dear @tdonohue @artlowel 
I am sorry, but I have to say we followed your ideas / instructions in this feature, and I think it was ready in Nov.
I know this is low priority at the moment, but from that time we have fixed several merge conflicts because the main branch is changing and changing. 
I know this PR got low priority, so now I have to say I told my developer @dsipos-dev to wait until somebody of your team can deal with it. (Because previously we have fixed the merge conflict immediately it got the merge conflict label.)

**Review by tdonohue — COMMENTED** [review]
@dsipos-dev and @kanasznagyzoltan : My apologies for waiting so long to look at this.  I wanted to be sure that we got Angular upgraded *first* (to v10) before adding lower priority features in.

I downloaded this PR today and gave it a test, and I'm currently **not** able to get this working.  I'm not sure if I did something wrong, or if this library is not working fully with Angular 10.  Here's the steps I took:

1. Submitted a new Item with 4 images attached to it (all JPG and PNG).
2. Ran the (backend) `./dspace filter-media` script manually to ensure a thumbnail was generated
3. Visited that Item page (http://localhost:4000/items/a9121346-0b6c-4f41-ba07-8d0c775f4925).  I see the thumbnail & I see the links to the 4 images, but **I do not see the media viewer.**  I tried clicking the thumbnail and nothing happens.
4. Then, I created a second item with a Video (MP4) attached.
5. Went to that Item's page.  On that page, I see the "No thumbnail" image (as we don't currently create thumbnails for videos), but it's not clickable.

It's possible I've misunderstood how to enable this or get this working.  Any advice on testing this?

As a sidenote, I see this PR is currently throwing some (minor) linting errors.  That's the reason for the red ❌ right now.

**Comment by dsipos-dev** [comment]
Dear @tdonohue 
If you are on the newest commit you just have to change the environment variable mediaViewer: { image: true, video: true, }, and itt will work properly

**Review by tdonohue — CHANGES_REQUESTED** [review]
@dsipos-dev : I tried again today and I *was* able to get this working properly in Angular 10.  Thanks for the notes, I had missed that it is turned "off" by default (and that's fine).  Overall, it looks quite nice & works well.

One minor thing is that the image viewer is not "obvious".  It just looks like the normal thumbnail and it was not clear (to me at least) that clicking the image would do anything. It worked great once I clicked it, but we may want to think about a way to make it more obvious that there's an embedded image viewer there.

I've added a few additional minor inline questions/comments below.  I'll also note that currently this PR is failing "lint" checks, which means the tests are not even running.  So, before this could move forward, we'd need to get this PR cleaned up.  This is the primary reason I've submitted this review as a "Request changes", as the code needs to be updated to pass lint & tests. 

All that said, honestly, this is looking good to me!  I appreciate your patience in keeping this PR up-to-date as we've been in active development.  Personally, though, I think this PR is looking like something we could accept for the upcoming 7.0 Beta 5 (and final) release, assuming no other team members / reviewers have major concerns here.  Thanks again!

**Comment by kanasznagyzoltan** [comment]
Dear @tdonohue,
Thank you for your feedback. I will discuss those points with my developer @dsipos-dev.
We are glad, this extension could be useful for the upcoming DSpace 7.
How much time we have to complete and fix your comments? We would like this addition to be in the upcoming beta 5.
Thank you!
Kind regards,
Zoli.

**Comment by dsipos-dev** [comment]
Dear @tdonohue 
I added your requests to the merge request and fix the build failing.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @dsipos-dev and @kanasznagyzoltan .  This is looking good to me & all my prior feedback has been addressed.  I've also retested and it works well.

(As a sidenote, I'm even debating whether we should default these viewers to both be *enabled* by default, as I suspect most will want these enabled.  That said, I'll wait for feedback from others on that.)

@artlowel or @atarix83 : Could one of you give this a second look to verify you still approve?

**Comment by kanasznagyzoltan** [comment]
Thank you @tdonohue! Thank you for your feedback!
Is everything OK? Now our solution is ready to be included in the beta5?

**Comment by tdonohue** [comment]
@kanasznagyzoltan : Per our development policies, all new features require *two approvals* before they can be merged. This is simply waiting on a second approval, likely from either @artlowel or @atarix83 .  I'll touch base with them in this week's meeting.  So, there's nothing else to do on your end.

**Review by artlowel — APPROVED** [review]
Thanks @dsipos-dev!

As far as I'm concerned this can be merged as soon as the merge conflicts are fixed.

**Comment by tdonohue** [comment]
@dsipos-dev and/or @kanasznagyzoltan : Could one of you find time to resolve the merge conflicts in the next week or so?  This PR now has 2 approvals, so it can be merged once the merge conflicts are resolved.  Thanks!

**Comment by kanasznagyzoltan** [comment]
Dear @tdonohue, next week we will definitely have time to resolve the conflicts, please wait us! :) Thank you.

File: src/styles/_global-styles.scss
Comment (tdonohue) [inline src/styles/_global-styles.scss:39]: This change causes the breadcrumbs bar to go over top of the login dropdown.  Here's a screenshot (notice how the login button is hidden behind the gray bar):
login-overlayed

At least locally, it looks like reverting this change solves the problem & the media viewer still appears to work

**Review by tdonohue — COMMENTED** [review]
@dsipos-dev : I gave this one final test before merging, and I stumbled on a very small bug.  Would you be able to quickly fix this?  If not, I could merge and fix in a follow-up PR.  Thanks!

File: src/styles/_global-styles.scss
Comment (dsipos-dev) [reply src/styles/_global-styles.scss:39]: If i revert this row the header is appears over the image media viewer. The header will block the action buttons  
image

**Comment by dsipos-dev** [comment]
@tdonohue I fixed the css problem with the login bar

**Comment by tdonohue** [comment]
@dsipos-dev : Thank you!!!  I've retested today and the issues are now resolved.  Merging this now as it's at +2.  Thanks again.

**Pull request merged**
